### PR TITLE
Clear old Validations during online delete (RIPD-870):

### PR DESCRIPTION
**Pull request by ximinez**
* Add Validations.LedgerSeq and .InitialSeq fields.
* Clean up logging.
* Lower online delete minimum for standalone mode.
* Unit tests of online_delete.

This is a second attempt. The first attempt was #1525.

My "large data set" testing procedure started with the existing installation on my dogfood box, running as a validator. Notably, ledger.db had over 8,000,000 rows in the Validations table (unfortunately, not as many as previously, because I'm a dummy and didn't save that DB). I configured online_delete=150000 and waited for online delete to kick in and finish cleaning up the Validators table.

The cleanup deleted 8,221,110 rows in 2 hours 23 minutes - an average of 960 per second - and did it in one shot. The state never changed, sync was never lost, everything stayed happy.

Reviewers: @mtrippled @seelabs

### CODE REVIEW

**Comment by mtrippled** [comment]
:+1:

File: src/ripple/app/ledger/Ledger.cpp [hunk 1]
```diff
@@ -1074,18 +1064,51 @@ static bool saveValidatedLedger (
     }
 
     {
+        static std::string addLedger(
+            "INSERT OR REPLACE INTO Ledgers "
+            "(LedgerHash,LedgerSeq,PrevHash,TotalCoins,ClosingTime,PrevClosingTime,"
+            "CloseTimeRes,CloseFlags,AccountSetHash,TransSetHash) VALUES "
+            "(:ledgerHash,:ledgerSeq,:prevHash,:totalCoins,:closingTime,:prevClosingTime,"
+            ":closeTimeRes,:closeFlags,:accountSetHash,:transSetHash);");
```
Comment (seelabs) [inline src/ripple/app/ledger/Ledger.cpp]: Nit: I like raw strings for long sql statements like this.

File: src/ripple/app/ledger/Ledger.cpp [hunk 2]
```diff
@@ -1074,18 +1064,51 @@ static bool saveValidatedLedger (
...
+
+        *db << addLedger,
+            soci::use(hash),
+            soci::use(seq),
+            soci::use(parentHash),
+            soci::use(drops),
+            soci::use(closeTime),
+            soci::use(parentCloseTime),
+            soci::use(closeTimeResolution),
+            soci::use(ledger->info().closeFlags),
```
Comment (seelabs) [inline src/ripple/app/ledger/Ledger.cpp]: Nit: This works. However, we avoid function calls within `soci::use` because of the dangers of taking the addresses of temps (even though it's OK in this case). I'd prefer if function calls are banned from `soci::use`, it will save time from always having to check that `info()` returns a ref.

File: src/ripple/app/main/DBInit.cpp [hunk 3]
```diff
@@ -90,14 +90,29 @@ const char* LedgerDBInit[] =
...
         SignTime    BIGINT UNSIGNED,                \
         RawData     BLOB                            \
     );",
+    // This will error out if the column already exists,
+    //  but DatabaseCon intentionally ignores errors.
+    "ALTER TABLE Validations                        \
+        ADD COLUMN LedgerSeq       BIGINT UNSIGNED;",
+    "ALTER TABLE Validations                        \
+        ADD COLUMN InitialSeq      BIGINT UNSIGNED;",
+
```
Comment (seelabs) [inline src/ripple/app/main/DBInit.cpp]: Can we write this so we only alter the table if the column doesn't exist? If it's hard, don't worry about it.

File: src/ripple/app/misc/SHAMapStoreImp.cpp [hunk 4]
```diff
@@ -593,6 +586,119 @@ SHAMapStoreImp::clearPrior (LedgerIndex lastRotated)
...
+                R"sql(DELETE FROM Validations
+                WHERE LedgerHash IN
+                (
+                    SELECT v.LedgerHash
+                    FROM Validations v
+                    WHERE v.LedgerSeq is NULL AND
+                        (v.InitialSeq IS NULL OR v.InitialSeq < )sql" +
+                    std::to_string(lastRotated) +
+                    R"sql()
+                    LIMIT )sql" +
```
Comment (seelabs) [inline src/ripple/app/misc/SHAMapStoreImp.cpp]: Nit: I think in this case the raw string adds more clutter than it's worth. I'd just make a regular string on one line.

File: src/ripple/core/impl/DatabaseCon.cpp [hunk 5]
```diff
@@ -46,7 +46,9 @@ DatabaseCon::DatabaseCon (
     {
         try
         {
-            session_ << initStrings[i];
+            soci::statement st = session_.prepare <<
+                initStrings[i];
+            st.execute(true);
```
Comment (seelabs) [inline src/ripple/core/impl/DatabaseCon.cpp]: Nit: This works, but why the change? I don't see what's wrong with what was there originally.

**Comment by seelabs** [comment]
:+1: I only found nits.

File: src/ripple/app/ledger/Ledger.cpp (hunk 2 above)
Comment (ximinez) [inline src/ripple/app/ledger/Ledger.cpp]: In retrospect, it's not the best decision to rely on the implementation of ledger and info to be safe for soci, so I converted this to use another local.

File: src/ripple/app/main/DBInit.cpp (hunk 3 above)
Comment (ximinez) [inline src/ripple/app/main/DBInit.cpp]: I did a lot of digging into this, and as far as I can find, it's completely impossible with sqlite's SQL, and doing it in C++ would (a) require reimplementing this to use more complicated logic than just a array of strings, and (2) the implementation would be quite a bit more complicated than "ignore SQL errors".

File: src/ripple/app/misc/SHAMapStoreImp.cpp (hunk 4 above)
Comment (ximinez) [inline src/ripple/app/misc/SHAMapStoreImp.cpp]: Agreed. It's a lot more readable this way.

File: src/ripple/core/impl/DatabaseCon.cpp (hunk 5 above)
Comment (ximinez) [inline src/ripple/core/impl/DatabaseCon.cpp]: Because what was there originally _should_ have worked, but actually had some internal temp sqlite object get an invalid reference count and call `terminate()`. Put another way, errors were not actually ignored. I'm reasonably sure it's a bug in sqllite, but I couldn't find it, so I used a workaround that also worked.

**Comment by ximinez** [comment]
Addressed @seelabs's nits, and rebased onto 0.31.0-b11 to resolve conflicts.

**Comment by ximinez** [comment]
I'm gonna go ahead and merk this as passed.

File: src/ripple/app/main/DBInit.cpp (hunk 3 above)
Comment (JoelKatz) [inline src/ripple/app/main/DBInit.cpp]: Unless I'm missing something, we used to do this all the time when we had database changes. We'd query the structure of the database and if it was missing a column we needed, we'd add that column.

Comment (mtrippled) [inline src/ripple/app/main/DBInit.cpp]: If you select a column that isn't there, sqlite gives an error about a not there column.

[rippled@bm2 db]$ sqlite3 ledger.db
SQLite version 3.7.17 2013-05-20 00:56:22
Enter ".help" for instructions
Enter SQL statements terminated with a ";"
sqlite> select ledgerseq from ledgers limit 10;
19347261
19347262
19347263
19347264
19347265
19347266
19347267
19347268
19347269
19347270
sqlite> select foo from ledgers;
Error: no such column: foo

Comment (mtrippled) [inline src/ripple/app/main/DBInit.cpp]: Though adding a column that is already there also results in an error and does not add a duplicate column.

sqlite> alter table ledgers add column ledgerseq;
Error: duplicate column name: ledgerseq

Comment (ximinez) [inline src/ripple/app/main/DBInit.cpp]: @JoelKatz That's exactly the kind of functionality I would have reused if I had found it the first time around. I'll update to do the database changes the right way.
